### PR TITLE
Skale-contracts integration

### LINKED ISSUE #1005: Prepare infrastructure for skale-contracts

### LINKED ISSUE #1006: Integrate skale-contracts into the upgrade script
```[tasklist]
### Tasks
- [ ] https://github.com/skalenetwork/skale-manager/issues/1005
- [ ] https://github.com/skalenetwork/upgrade-tools/issues/222
- [ ] https://github.com/skalenetwork/skale-manager/issues/1062
```

### PR DESCRIPTION
**Pull request by DimaStebaev**
Add skale-contracts support
Replace `ABI` parameter with `TARGET`. Where `TARGET` is an alias or address of the upgraded instance.
Move to upgrade-tools 3
Move to ethers v6
Fix discrepancy in gas reimbursement
Fix errors in tests

### CODE REVIEW

File: test/SkaleDKG.ts [hunk 1]
```diff
@@ -2674,10 +2680,10 @@ describe("SkaleDKG", () => {
                 //     continue;
                 // }
```
Comment (DmytroNazarenko) [inline test/SkaleDKG.ts]: Remove commented code, please

File: test/utils/SegmentTree.ts [hunk 2]
```diff
@@ -119,84 +117,84 @@ describe("SegmentTree", () => {
                 //     }
                 //     console.log(str);
                 // }
```
Comment (DmytroNazarenko) [inline test/utils/SegmentTree.ts]: Remove commented code, please

File: test/utils/SegmentTree.ts [hunk 3]
```diff
@@ -348,20 +342,20 @@ describe("SegmentTree", () => {
                 await segmentTree.addToPlace(106, 25);
                 await segmentTree.addToPlace(77, 509);
                 for(let i = 0; i < 180; i++) {
-                    const place = (await segmentTree.getRandomElem(78)).toNumber();
+                    const place = (await segmentTree.getRandomElem(78));
                     await segmentTree.removeFromPlace(place, 1);
                     // console.log("Place found!!!!!!!!!!!!")
                     // console.log(place);
```
Comment (DmytroNazarenko) [inline test/utils/SegmentTree.ts]: Remove commented code, please

File: test/utils/SegmentTree.ts [hunk 4]
```diff
@@ -348,20 +342,20 @@ describe("SegmentTree", () => {
...
             });
 
             it("random stress simulating large schains test", async () => {
-                const schainPlace = 32; // 1/4 of node
+                const schainPlace = 32n; // 1/4 of node
                 await segmentTree.removeFromPlace(128, 100); // make 50 nodes
                 for(let i = 0; i < 200; i++) { // 200 times we could repeat removing
-                    const place = (await segmentTree.getRandomElem(schainPlace)).toNumber();
+                    const place = (await segmentTree.getRandomElem(schainPlace));
                     // console.log("New place ", place);
```
Comment (DmytroNazarenko) [inline test/utils/SegmentTree.ts]: Remove, please

File: test/utils/SegmentTree.ts [hunk 5]
```diff
@@ -370,14 +364,14 @@ describe("SegmentTree", () => {
...
             it("random stress simulating large schains test moving elements", async () => {
-                const schainPlace = 32; // 1/4 of node
+                const schainPlace = 32n; // 1/4 of node
                 await segmentTree.removeFromPlace(128, 100); // make 50 nodes
                 for(let i = 0; i < 200; i++) { // 200 times we could repeat removing
-                    const place = (await segmentTree.getRandomElem(schainPlace)).toNumber();
+                    const place = (await segmentTree.getRandomElem(schainPlace));
                     // console.log("New place ", place);
                     if (place - schainPlace > 0) {
                         // console.log(place - schainPlace);
```
Comment (DmytroNazarenko) [inline test/utils/SegmentTree.ts]: Remove please

File: test/SkaleDKG.ts (hunk 1 above)
Comment (DimaStebaev) [reply test/SkaleDKG.ts]: Done.

File: test/utils/SegmentTree.ts (hunk 2 above)
Comment (DimaStebaev) [reply test/utils/SegmentTree.ts]: Done.

File: test/utils/SegmentTree.ts (hunk 3 above)
Comment (DimaStebaev) [reply test/utils/SegmentTree.ts]: Done.

File: test/utils/SegmentTree.ts (hunk 4 above)
Comment (DimaStebaev) [reply test/utils/SegmentTree.ts]: Done.

File: test/utils/SegmentTree.ts (hunk 5 above)
Comment (DimaStebaev) [reply test/utils/SegmentTree.ts]: Done.

**Review by DmytroNazarenko — APPROVED** [review]

**Pull request merged**